### PR TITLE
issue60-binary-projects

### PR DESCRIPTION
**Pull request by DannyPeck**
- Hatch is now capable of generating binary project versions of its assets.
- Added more robust test cases.
- Cleaned up asset tests.
- Switch Arch::X32 to Arch::X86.

### CODE REVIEW

File: hatch/src/assets/tests/generator.rs [hunk 1]
```diff
@@ -8,16 +8,22 @@ use std::path::PathBuf;
 fn generate_one_without_directories() {
   let test_asset = ProjectAsset::new(PathBuf::from("./"), String::from("test.test"), String::from("test"));
 
-  generate_one(&test_asset);
+  if let Err(e) = generate_one(&test_asset) {
+    panic!(e);
```
Comment (technetos) [inline hatch/src/assets/tests/generator.rs:12]: Im not super happy with the use of panic!

Is there a better way to fail here?

File: hatch/src/assets/tests/tuprules.rs [hunk 2]
```diff
@@ -1,13 +1,183 @@
...
+/**
+ * These tests were created using All-Pairs testing.
+ *
+ * Note that I am not including the Compiler field as it is required and we are simply doing a
+ * string replace as opposed to an enumeration.
+ *
+ * the tests are following this table:
+ * (Kind,   Arch,  Target, Compiler Flags, Linker Flags)
+ * (Static,  X64, Release,        Present,      Present)
+ * (Static,  X86,   Debug,           None,         None)
```
Comment (technetos) [inline hatch/src/assets/tests/tuprules.rs:13]: We are not testing Static + X64 + Debug why?

**Review by technetos — CHANGES_REQUESTED** [review]

File: hatch/src/assets/tests/generator.rs (hunk 1 above)
Comment (DannyPeck) [reply hatch/src/assets/tests/generator.rs:12]: This is how Rust tests work, they detect failure via panic.

File: hatch/src/assets/tests/tuprules.rs (hunk 2 above)
Comment (DannyPeck) [reply hatch/src/assets/tests/tuprules.rs:13]: This set of test cases was determined using a testing technique called All-Pairs testing.  The goal of All-Pairs testing is to reduce the number of test cases while maintaining the same level of confidence.  All-Pairs is centered around the idea that each value is paired with every other value at least once.  Example: in the first test case Static is paired with X64, Release, Present, and Present.  X64 is paired with Static, Release, Present, and Present. Release is paired with Static, X64, Present, and Present.  So on and so on.  The point being that I have no satisfied those pairing and do not have to test those pairing again.  You'll see in the next test case that I am no matching Static with X86, Debug, None, and None.  At  this point Static has been paired with all possible values for the other 4 parameters and is complete.  X64 and X86 have been paired with Static but are still missing being paired with Shared and Binary.  This process continues until all pairs have been accounted for.  What this approach allows you to do is cover more conditions in less test cases.  I am actually missing some test cases here, I have forgot to test the pair combinations of Compiler Flags: Present, and Linker Flags: None as well as Compiler Flags: None, and Linker Flags: Present.

I hope this helps!

File: hatch/src/assets/tests/generator.rs (hunk 1 above)
Comment (technetos) [reply hatch/src/assets/tests/generator.rs:12]: Ok

File: hatch/src/assets/tests/tuprules.rs (hunk 2 above)
Comment (technetos) [reply hatch/src/assets/tests/tuprules.rs:13]: That's much clearer thank-you for the explanation.

**Comment by DannyPeck** [comment]
I have fixed the failing test and added the two missing All-Pairs tests.

**Review by technetos — DISMISSED** [review]

**Review by technetos — APPROVED** [review]
